File: src/dotnet/Src/ButterflyNet.Parser/Strategies/UnparsedStrategy.cs
using System.Text.RegularExpressions;
using ButterflyNet.Parser.Scopes;

namespace ButterflyNet.Parser.Strategies {
	[TokenTransformer("[!")]
	public class UnparsedStrategy : ScopeDrivenStrategy {
		private static readonly Regex unparsedRegex = new Regex(@"(.*?)](?!])", RegexOptions.Singleline);
		
		public override int Priority { get { return DefaultPriority - 1; } }

		protected override void DoExecute(ParseContext context) {
			OpenScope(new UnparsedScope(), context);

			var match = unparsedRegex.Match(context.Input.PeekSubstring);
			if (!match.Success) {
				throw new ParseException("Unparsed scope never closes");
			}

			var text = match.Groups[1].Value.Replace("]]", "]");
			context.Input.Read(match.Value.Length);
			context.UpdateCurrentChar();
			context.Analyzer.WriteAndEscape(text);
			CloseCurrentScope(context);
		}
	}
}

File: src/js/src/strategies/UnparsedStrategy.js
function UnparsedStrategy() {
	UnparsedStrategy.$parent.call(this);

	this.setAsTokenTransformer("[!");
	this.priority = ParseStrategy.defaultPriority - 1;

	this.doExecute = function(context) {
		this.openScope(new UnparsedScope(), context);
		
		var match = /([\s\S]*?)](?!])/.exec(context.input.peekSubstring());
		if (!match) {
			throw new ParseException("Unparsed scope never closes");
		}
		
		var text = match[1].replace(/]]/g, "]");
		context.input.read(match[0].length);
		context.updateCurrentChar();
		context.analyzer.writeAndEscape(text);
		this.closeCurrentScope(context);
	};
}

extend(ScopeDrivenStrategy, UnparsedStrategy);